File: Python/services/student_service.py
"""handles database stuff for students"""

from typing import List, Optional, Dict, Any, Tuple

from database.db_handler import DatabaseHandler
from models.student import Student


class StudentService:
    """does database operations for students"""

    def __init__(self, db_handler: DatabaseHandler):
        """setup with database connection"""
        self.db_handler = db_handler

    def add_student(self, student: Student) -> Student:
        """add new student to database"""
        # check course format
        if not Student.validate_course(student.course):
            raise ValueError(
                f"Bad course format: {student.course}. Should be like CS101.")

        # check duplicates
        if self.is_duplicate_name(student.name):
            raise ValueError(
                f"Name '{student.name}' already exists. Try adding initials.")

        query = """
        INSERT INTO students (name, course)
        VALUES (?, ?);
        """
        params = (student.name, student.course)

        student_id = self.db_handler.execute_write_query(query, params)
        student.student_id = student_id

        return self.get_student_by_id(student_id)

    def get_student_by_id(self, student_id: int) -> Optional[Student]:
        """get one student by ID"""
        query = "SELECT * FROM students WHERE student_id = ?;"
        result = self.db_handler.execute_query(query, (student_id,))

        if not result:
            return None

        return Student.from_db_dict(result[0])

    def get_all_students(self) -> List[Student]:
        """get all students"""
        query = "SELECT * FROM students ORDER BY student_id;"
        results = self.db_handler.execute_query(query)

        return [Student.from_db_dict(row) for row in results]

    def update_student(self, student: Student) -> Optional[Student]:
        """update student info"""
        if student.student_id is None:
            raise ValueError("Can't update without ID")

        # check course
        if not Student.validate_course(student.course):
            raise ValueError(
                f"Bad course format: {student.course}. Should be like CS101.")

        # check duplicate name
        if self.is_duplicate_name(student.name, student.student_id):
            raise ValueError(
                f"Name '{student.name}' already exists. Try adding initials.")

        query = """
        UPDATE students
        SET name = ?, course = ?
        WHERE student_id = ?;
        """
        params = (student.name, student.course, student.student_id)

        affected_rows = self.db_handler.execute_write_query(query, params)
        if affected_rows == 0:
            return None

        return self.get_student_by_id(student.student_id)

    def update_students_course(self, student_ids: List[int], new_course: str) -> int:
        """update course for multiple students"""
        # check course format
        if not Student.validate_course(new_course):
            raise ValueError(
                f"Bad course format: {new_course}. Should be like CS101.")

        # convert ids to sql format
        id_placeholder = ','.join('?' for _ in student_ids)

        query = f"""
        UPDATE students
        SET course = ?
        WHERE student_id IN ({id_placeholder});
        """

        params = (new_course,) + tuple(student_ids)

        return self.db_handler.execute_write_query(query, params)

    def delete_student(self, student_id: int) -> bool:
        """delete a student"""
        query = "DELETE FROM students WHERE student_id = ?;"
        affected_rows = self.db_handler.execute_write_query(
            query, (student_id,))

        return affected_rows > 0

    def delete_students(self, student_ids: List[int]) -> int:
        """delete multiple students"""
        id_placeholder = ','.join('?' for _ in student_ids)

        query = f"""
        DELETE FROM students
        WHERE student_id IN ({id_placeholder});
        """

        return self.db_handler.execute_write_query(query, tuple(student_ids))

    def search_students(self, search_term: str) -> List[Student]:
        """search by name or course"""
        query = """
        SELECT * FROM students
        WHERE name LIKE ? OR course LIKE ?
        ORDER BY name;
        """
        search_pattern = f"%{search_term}%"
        params = (search_pattern, search_pattern)

        results = self.db_handler.execute_query(query, params)

        return [Student.from_db_dict(row) for row in results]

    def advanced_search(self, criteria: Dict[str, Any]) -> List[Student]:
        """search with filters"""
        params = []
        conditions = []

        # build query
        if 'student_id' in criteria and criteria['student_id']:
            conditions.append("student_id = ?")
            params.append(criteria['student_id'])

        if 'name' in criteria and criteria['name']:
            conditions.append("name LIKE ?")
            params.append(f"%{criteria['name']}%")

        if 'course' in criteria and criteria['course']:
            conditions.append("course LIKE ?")
            params.append(f"%{criteria['course']}%")

        if 'enrollment_date_from' in criteria and criteria['enrollment_date_from']:
            conditions.append("enrollment_date >= ?")
            params.append(criteria['enrollment_date_from'])

        if 'enrollment_date_to' in criteria and criteria['enrollment_date_to']:
            conditions.append("enrollment_date <= ?")
            params.append(criteria['enrollment_date_to'])

        # if no conditions, get everything
        if not conditions:
            return self.get_all_students()

        query = f"""
        SELECT * FROM students
        WHERE {' AND '.join(conditions)}
        ORDER BY name;
        """

        results = self.db_handler.execute_query(query, tuple(params))

        return [Student.from_db_dict(row) for row in results]

    def is_duplicate_name(self, name: str, exclude_id: Optional[int] = None) -> bool:
        """check if name already exists"""
        if exclude_id:
            query = "SELECT COUNT(*) as count FROM students WHERE name = ? AND student_id != ?;"
            result = self.db_handler.execute_query(query, (name, exclude_id))
        else:
            query = "SELECT COUNT(*) as count FROM students WHERE name = ?;"
            result = self.db_handler.execute_query(query, (name,))

        return result[0]['count'] > 0

File: java/src/main/java/com/studentmanagement/services/StudentService.java
package com.studentmanagement.services;

import com.studentmanagement.database.DatabaseHandler;
import com.studentmanagement.models.Student;
import java.sql.SQLException;
import java.util.*;
import java.util.stream.Collectors;
import java.lang.Runtime;

public class StudentService {
    private final DatabaseHandler dbHandler;

    public StudentService(DatabaseHandler dbHandler) {
        this.dbHandler = dbHandler;
    }

    // profiling helper
    private void logProfile(String methodName, long startTime, long startMem) {
        long endTime = System.nanoTime();
        long endMem = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory();
        System.out.printf("[PROFILE] StudentService.%s: time=%.3fs; memDelta=%.2fKB%n", methodName, (endTime - startTime)/1e9, (endMem - startMem)/1024.0);
    }

    public Student addStudent(Student student) throws SQLException {
        long startTime = System.nanoTime();
        long startMem = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory();
        if (!Student.validateCourse(student.getCourse())) {
            throw new IllegalArgumentException("Bad course format: " + student.getCourse());
        }
        if (isDuplicateName(student.getName(), null)) {
            throw new IllegalArgumentException("Name '" + student.getName() + "' already exists.");
        }
        String sql = "INSERT INTO students (name, course) VALUES (?, ?);";
        int id = dbHandler.executeUpdate(sql, student.getName(), student.getCourse());
        Student result = getStudentById(id);
        logProfile("addStudent", startTime, startMem);
        return result;
    }

    public Student getStudentById(int studentId) throws SQLException {
        long startTime = System.nanoTime();
        long startMem = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory();
        String sql = "SELECT * FROM students WHERE student_id = ?;";
        List<Map<String, Object>> rows = dbHandler.executeQuery(sql, studentId);
        Student result = rows.isEmpty() ? null : Student.fromMap(rows.get(0));
        logProfile("getStudentById", startTime, startMem);
        return result;
    }

    public List<Student> getAllStudents() throws SQLException {
        long startTime = System.nanoTime();
        long startMem = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory();
        String sql = "SELECT * FROM students ORDER BY student_id;";
        List<Map<String, Object>> rows = dbHandler.executeQuery(sql);
        List<Student> result = rows.stream().map(Student::fromMap).collect(Collectors.toList());
        logProfile("getAllStudents", startTime, startMem);
        return result;
    }

    public Student updateStudent(Student student) throws SQLException {
        long startTime = System.nanoTime();
        long startMem = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory();
        if (student.getStudentId() == null) {
            throw new IllegalArgumentException("Can't update without ID");
        }
        if (!Student.validateCourse(student.getCourse())) {
            throw new IllegalArgumentException("Bad course format: " + student.getCourse());
        }
        if (isDuplicateName(student.getName(), student.getStudentId())) {
            throw new IllegalArgumentException("Name '" + student.getName() + "' already exists.");
        }
        String sql = "UPDATE students SET name = ?, course = ? WHERE student_id = ?;";
        int affected = dbHandler.executeUpdate(sql, student.getName(), student.getCourse(), student.getStudentId());
        Student result = affected == 0 ? null : getStudentById(student.getStudentId());
        logProfile("updateStudent", startTime, startMem);
        return result;
    }

    public int updateStudentsCourse(List<Integer> studentIds, String newCourse) throws SQLException {
        long startTime = System.nanoTime();
        long startMem = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory();
        if (!Student.validateCourse(newCourse)) {
            throw new IllegalArgumentException("Bad course format: " + newCourse);
        }
        if (studentIds == null || studentIds.isEmpty()) return 0;
        String placeholders = studentIds.stream().map(id -> "?").collect(Collectors.joining(","));
        String sql = String.format("UPDATE students SET course = ? WHERE student_id IN (%s);", placeholders);
        List<Object> params = new ArrayList<>();
        params.add(newCourse);
        params.addAll(studentIds);
        int result = dbHandler.executeUpdate(sql, params.toArray());
        logProfile("updateStudentsCourse", startTime, startMem);
        return result;
    }

    public boolean deleteStudent(int studentId) throws SQLException {
        long startTime = System.nanoTime();
        long startMem = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory();
        String sql = "DELETE FROM students WHERE student_id = ?;";
        int affected = dbHandler.executeUpdate(sql, studentId);
        boolean result = affected > 0;
        logProfile("deleteStudent", startTime, startMem);
        return result;
    }

    public int deleteStudents(List<Integer> studentIds) throws SQLException {
        long startTime = System.nanoTime();
        long startMem = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory();
        if (studentIds == null || studentIds.isEmpty()) return 0;
        String placeholders = studentIds.stream().map(id -> "?").collect(Collectors.joining(","));
        String sql = String.format("DELETE FROM students WHERE student_id IN (%s);", placeholders);
        int result = dbHandler.executeUpdate(sql, studentIds.toArray());
        logProfile("deleteStudents", startTime, startMem);
        return result;
    }

    public List<Student> searchStudents(String searchTerm) throws SQLException {
        long startTime = System.nanoTime();
        long startMem = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory();
        String sql = "SELECT * FROM students WHERE name LIKE ? OR course LIKE ? ORDER BY name;";
        String pattern = "%" + searchTerm + "%";
        List<Map<String, Object>> rows = dbHandler.executeQuery(sql, pattern, pattern);
        List<Student> result = rows.stream().map(Student::fromMap).collect(Collectors.toList());
        logProfile("searchStudents", startTime, startMem);
        return result;
    }

    public List<Student> advancedSearch(Map<String, Object> criteria) throws SQLException {
        long startTime = System.nanoTime();
        long startMem = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory();
        List<String> conditions = new ArrayList<>();
        List<Object> params = new ArrayList<>();

        if (criteria.containsKey("student_id")) {
            conditions.add("student_id = ?");
            params.add(criteria.get("student_id"));
        }
        if (criteria.containsKey("name")) {
            conditions.add("name LIKE ?");
            params.add("%" + criteria.get("name") + "%");
        }
        if (criteria.containsKey("course")) {
            conditions.add("course LIKE ?");
            params.add("%" + criteria.get("course") + "%");
        }
        if (criteria.containsKey("enrollment_date_from")) {
            conditions.add("enrollment_date >= ?");
            params.add(criteria.get("enrollment_date_from"));
        }
        if (criteria.containsKey("enrollment_date_to")) {
            conditions.add("enrollment_date <= ?");
            params.add(criteria.get("enrollment_date_to"));
        }
        if (conditions.isEmpty()) {
            return getAllStudents();
        }
        String whereClause = String.join(" AND ", conditions);
        String sql = String.format("SELECT * FROM students WHERE %s ORDER BY name;", whereClause);
        List<Map<String, Object>> rows = dbHandler.executeQuery(sql, params.toArray());
        List<Student> result = rows.stream().map(Student::fromMap).collect(Collectors.toList());
        logProfile("advancedSearch", startTime, startMem);
        return result;
    }

    public boolean isDuplicateName(String name, Integer excludeId) throws SQLException {
        long startTime = System.nanoTime();
        long startMem = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory();
        String sql;
        List<Map<String, Object>> result;
        if (excludeId != null) {
            sql = "SELECT COUNT(*) as count FROM students WHERE name = ? AND student_id != ?;";
            result = dbHandler.executeQuery(sql, name, excludeId);
        } else {
            sql = "SELECT COUNT(*) as count FROM students WHERE name = ?;";
            result = dbHandler.executeQuery(sql, name);
        }
        int count = ((Number) result.get(0).get("count")).intValue();
        boolean dup = count > 0;
        logProfile("isDuplicateName", startTime, startMem);
        return dup;
    }
} 